kedro/io/data_catalog.py

```diff
@@ -262,30 +266,143 @@ class to be loaded is specified with the key ``type`` and their
         save_version = save_version or generate_timestamp()
         load_versions = copy.deepcopy(load_versions) or {}
 
-        missing_keys = load_versions.keys() - catalog.keys()
+        layers: dict[str, set[str]] = defaultdict(set)
+
+        missing_keys = [
+            key
+            for key in load_versions.keys()
+            if not cls._match_name_against_pattern(catalog, key)  # type: ignore
+        ]
         if missing_keys:
             raise DatasetNotFoundError(
                 f"'load_versions' keys [{', '.join(sorted(missing_keys))}] "
                 f"are not found in the catalog."
             )
 
-        layers: dict[str, set[str]] = defaultdict(set)
-        for ds_name, ds_config in catalog.items():
-            ds_layer = ds_config.pop("layer", None)
-            if ds_layer is not None:
-                layers[ds_layer].add(ds_name)
-
-            ds_config = _resolve_credentials(ds_config, credentials)
-            data_sets[ds_name] = AbstractDataSet.from_config(
-                ds_name, ds_config, load_versions.get(ds_name), save_version
+        # Add load_version for dataset in raw catalog dict, create an entry if it
+        # doesn't exist because this might be a patterned dataset
+        for ds_name, version in load_versions.items():
+            catalog.setdefault(ds_name, {}).update({"load_version": version})
+        # Update raw catalog dict with save version and resolved credentials
+        catalog = {
+            ds_name: _resolve_credentials(
+                {**ds_config, "save_version": save_version}, credentials
             )
+            for ds_name, ds_config in catalog.items()
+        }
+        # Already add explicit entry datasets
+        for ds_name, ds_config in catalog.items():
+            if "}" not in ds_name and cls._is_full_config(ds_config):
```

Comment:
I think we need a small helper function here i.e. `_is_pattern` even if it's a one-liner, as the semantics make it easier to read.